include/dali/core/mm/async_pool.h

```diff
@@ -0,0 +1,428 @@
+   * (3) If (1) fails and there are free pending per-stream operations, an attempt is made to get
+   *     memory from the free list of the underlying global pool, without using upstream resource.
+   *     If this allocation succeeds, the memory allocated this way is returned.
+   * (4) If (3) fails, the pending per-stream frees are collected to the global pool and then
+   *     an allocation is made from the global pool.
+   *
+   * Per-stream frees are not coalesced - therefore, even if the total memory freed on a stream
+   * may be sufficient, there may be no satisfactory block.
+   */
+  void *do_allocate_async(size_t bytes, size_t alignment, stream_view stream) override {
+    adjust_size_and_alignment(bytes, alignment);
+    std::lock_guard<LockType> guard(lock_);
+    auto it = stream_free_.find(stream.value());
+    void *ptr;
+    if (it != stream_free_.end()) {
+      ptr = try_allocate(it->second, bytes, alignment);
+      if (ptr)
+        return ptr;
+    }
+    return allocate_from_global_pool(bytes, alignment);
+  }
+
+  /**
+   * @brief Allocates from a global pool, possibly releasing per-stream memory to the global pool.
+   *
+   * If the global pool can't allocate enough memory without allocating upstream blocks,
+   * and there are some per-stream pending frees, the completed per-streem frees are returned
+   * to the global pool.
+   * Then, another request to the global pool is made, this time allowing allocation from upstream.
+   */
+  void *allocate_from_global_pool(size_t bytes, size_t alignment) {
+    void *ptr;
+    if (num_pending_frees_ == 0) {
+      ptr = global_pool_.allocate(bytes, alignment);
+      return ptr;
+    }
+    ptr = global_pool_.try_allocate_from_free(bytes, alignment);
+    if (ptr)
+      return ptr;
+    for (auto &kv : stream_free_)
+      free_pending(kv.second);
+    ptr = global_pool_.allocate(bytes, alignment);
+    return ptr;
+  }
+
+  void do_deallocate_async(void *mem, size_t bytes, size_t alignment, stream_view stream) override {
+    adjust_size_and_alignment(bytes, alignment);
+    std::lock_guard<LockType> guard(lock_);
+    char *ptr = static_cast<char*>(mem);
+    restore_original_params(ptr, bytes, alignment);
+    deallocate_async_impl(stream_free_[stream.value()], ptr, bytes, alignment, stream.value());
+  }
+
+  /**
+   * @brief Increases minimum alignment based on size and quantizes the size to a multiple
+   *        of the alignment.
+   *
+   * If the alignment requirements are low compared to the size, the alignment requirement is
+   * increased, which improves reusability of the block. Further, the block size is aligned
+   * to a next multiple of the (new) alignment, reducing the need for block splitting in case
+   * of multiple allocations of similarly-sized blocks.
+   */
+  static void adjust_size_and_alignment(size_t &size, size_t &alignment) {
+    if (size == 0)
+      return;
+    int l = ilog2(size);
+    size_t min_align = (1 << (l >> 1));  // 2^(l/2)
```

Comment:
`l` looks very alike to `1`, can you rename it to increase readibility?

Reply:
sure